pkg/reconciler/pubsub/reconciler.go

```diff
@@ -18,221 +18,145 @@ package pubsub
-		SubscriptionID: subscription,
-		Secret:         secret,
-		Owner:          owner,
-	})
-}
-
-func (c *PubSubBase) EnsureTopicExists(ctx context.Context, owner kmeta.OwnerRefable, secret corev1.SecretKeySelector, project, topic string) (ops.OpsJobStatus, error) {
-	return c.ensureTopicJob(ctx, operations.TopicArgs{
-		Image:     c.TopicOpsImage,
-		Action:    ops.ActionExists,
-		ProjectID: project,
-		TopicID:   topic,
-		Secret:    secret,
-		Owner:     owner,
-	})
-}
-
-func (c *PubSubBase) EnsureTopicCreated(ctx context.Context, owner kmeta.OwnerRefable, secret corev1.SecretKeySelector, project, topic string) (ops.OpsJobStatus, error) {
-	return c.ensureTopicJob(ctx, operations.TopicArgs{
-		Image:     c.TopicOpsImage,
-		Action:    ops.ActionCreate,
-		ProjectID: project,
-		TopicID:   topic,
-		Secret:    secret,
-		Owner:     owner,
-	})
-}
-
-func (c *PubSubBase) EnsureTopicDeleted(ctx context.Context, owner kmeta.OwnerRefable, secret corev1.SecretKeySelector, project, topic string) (ops.OpsJobStatus, error) {
-	return c.ensureTopicJob(ctx, operations.TopicArgs{
-		Image:     c.TopicOpsImage,
-		Action:    ops.ActionDelete,
-		ProjectID: project,
-		TopicID:   topic,
-		Secret:    secret,
-		Owner:     owner,
-	})
-}
-
-func (c *PubSubBase) ensureTopicJob(ctx context.Context, args operations.TopicArgs) (ops.OpsJobStatus, error) {
-	job, err := c.getJob(ctx, args.Owner.GetObjectMeta(), labels.SelectorFromSet(operations.TopicJobLabels(args.Owner, args.Action)))
-	// If the resource doesn't exist, we'll create it
-	if apierrs.IsNotFound(err) {
-		c.Logger.Debugw("Job not found, creating.")
+// ReconcilePubSub reconciles Topic / PullSubscription given a PubSubSpec.
+// Sets the following Conditions in the Status field appropriately:
+// "TopicReady", and "PullSubscriptionReady"
+// Also sets the following fields in the pubsubable.Status upon success
+// TopicID, ProjectID, and SinkURI
+func (psb *PubSubBase) ReconcilePubSub(ctx context.Context, pubsubable duck.PubSubable, topic, resourceGroup string) (*pubsubsourcev1alpha1.Topic, *pubsubsourcev1alpha1.PullSubscription, error) {
+	if pubsubable == nil {
+		return nil, nil, fmt.Errorf("nil pubsubable passed in")
+	}
+	namespace := pubsubable.GetObjectMeta().GetNamespace()
+	name := pubsubable.GetObjectMeta().GetName()
+	spec := pubsubable.PubSubSpec()
+	status := pubsubable.PubSubStatus()
 
-		job = operations.NewTopicOps(args)
+	topics := psb.pubsubClient.PubsubV1alpha1().Topics(namespace)
+	t, err := topics.Get(name, v1.GetOptions{})
 
-		job, err := c.KubeClientSet.BatchV1().Jobs(args.Owner.GetObjectMeta().GetNamespace()).Create(job)
-		if err != nil || job == nil {
-			c.Logger.Errorw("Failed to create Job.", zap.Error(err))
-			return ops.OpsJobCreateFailed, err
+	if err != nil {
+		if !apierrs.IsNotFound(err) {
+			logging.FromContext(ctx).Desugar().Error("Failed to get Topics", zap.Error(err))
+			return nil, nil, fmt.Errorf("failed to get Topics: %s", err.Error())
```

Comment:
A better verb now is `%w` https://golang.org/pkg/fmt/#Errorf
Similar for all other places.

Reply:
I'm getting this GoLand error: Unknown verb 'w' (%w).
Do you mind if we leave this for later?

Reply:
FYI `%w` is new grammer supported in Go 1.13.
Prow is already running all jobs for this repo with Go 1.13, so if you update you local Go version to 1.13, you will be able to make this change.

Reply:
hmm I have 1.13... will sync with @yolocs tomorrow... I think this one is the last comment, and we might be ready to get this in! Thanks @chizhg !

Reply:
Done, thanks @yolocs for the suggestion!